src/main/java/com/google/cloud/tools/ide/login/GoogleLoginState.java

```diff
@@ -428,104 +400,79 @@ public void simulateLoginStatusChange(boolean login) {
     uiFacade.notifyStatusIndicator();
   }
 
-  private void updateUserCredentials(GoogleTokenResponse tokenResponse) {
+  private void updateCredentials(GoogleTokenResponse tokenResponse) {
     refreshToken = tokenResponse.getRefreshToken();
     accessToken = tokenResponse.getAccessToken();
     oAuth2Credential = makeCredential();
-    accessTokenExpiryTime = System.currentTimeMillis() / 1000
-        + tokenResponse.getExpiresInSeconds().longValue();
+    accessTokenExpiryTime = System.currentTimeMillis() / 1000 + tokenResponse.getExpiresInSeconds();
     email = queryEmail();
-    saveCredentials();
+    persistCredentials();
     uiFacade.notifyStatusIndicator();
     notifyLoginStatusChange(true);
   }
 
   private void retrieveSavedCredentials() {
+    Preconditions.checkState(!isLoggedIn(), "Should be called only once in the constructor.");
 
     OAuthData savedAuthState = authDataStore.loadOAuthData();
 
-    // the stored email can be null in the case where the external browser
-    // was launched, because we can't extract the email from the external
-    // browser
     if (savedAuthState.getRefreshToken() == null || savedAuthState.getStoredScopes().isEmpty()) {
       authDataStore.clearStoredOAuthData();
       return;
     }
 
-    accessToken = savedAuthState.getAccessToken();
-    refreshToken = savedAuthState.getRefreshToken();
-    accessTokenExpiryTime = savedAuthState.getAccessTokenExpiryTime();
-    this.email = savedAuthState.getStoredEmail();
-
-    isLoggedIn = true;
-
     if (!oAuthScopes.equals(savedAuthState.getStoredScopes())) {
       loggerFacade.logWarning(
           "OAuth scope set for stored credentials no longer valid, logging out.");
       loggerFacade.logWarning(oAuthScopes + " vs. " + savedAuthState.getStoredScopes());
-      logOut(false);
+      authDataStore.clearStoredOAuthData();
+      return;
     }
 
+    accessToken = savedAuthState.getAccessToken();
+    refreshToken = savedAuthState.getRefreshToken();
+    accessTokenExpiryTime = savedAuthState.getAccessTokenExpiryTime();
+    email = savedAuthState.getStoredEmail();
     oAuth2Credential = makeCredential();
-  }
 
-  private boolean checkLoggedIn(String message) {
-    if (!isLoggedIn) {
-      if (!logIn(message)) {
-        return false;
-      }
-      uiFacade.notifyStatusIndicator();
-    }
-    return true;
+    isLoggedIn = true;
   }
 
   private void notifyLoginStatusChange(boolean login) {
     synchronized(listeners) {
       for (LoginListener listener : listeners) {
         try {
           listener.statusChanged(login);
-        } catch (Throwable t) {
-          loggerFacade.logError("Exception in LoginListener", t);
+        } catch (Throwable thrown) {
```

Comment:
Per Effective Java don't catch Throwable. What might actually be thrown here? Catch that.